python/hail/methods/statgen.py

```diff
@@ -1688,56 +1677,105 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+    scores : :class:`Table`, optional
+        Table keyed by a single field of type `:py:data:`.tstr` with values
+        including all column keys of ``mt``, and with a
+        ``scores`` field of type :class:`tarray` of `:py:data:`.tfloat64` with
+        all array values of the same positive length.
+        Exactly one of `k` and `scores` must be specified.
     min_kinship : :obj:`float`
         Pairs of samples with kinship lower than ``min_kinship`` are excluded
         from the results.
     statistics : :obj:`str`
-        the set of statistics to compute, `phi` will only compute the
-        kinship statistic, `phik2` will compute the kinship and
-        identity-by-descent two statistics, `phik2k0` will compute the
-        kinship statistics and both identity-by-descent two and zero, `all`
-        computes the kinship statistic and all three identity-by-descent
-        statistics.
+        Set of statistics to compute.
+        If ``'phi'``, only estimate the kinship statistic.
+        If ``'phik2'``, estimate the above and IBD2.
+        If ``'phik2k0'``, estimate the above and IBD0.
+        If ``'all'``, estimate the above and IBD1.
+    block_size : :obj:`int`, optional
+        Block size of block matrices used in the algorithm.
+        Default given by :meth:`.BlockMatrix.default_block_size`.
 
     Returns
     -------
     :class:`.Table`
-        A :class:`.Table` mapping pairs of samples to estimations of their
-        kinship and identity-by-descent zero, one, and two.
+        A :class:`.Table` mapping pairs of samples to their pair-wise statistics.
+    """
+    mt = require_biallelic(mt, 'pc_relate')
+    mt = require_unique_samples(mt, 'pc_relate')
+    require_col_key_str(mt, 'pc_relate')  # FIXME relax
+
+    assert len(mt.col_key) == 1
+    ck_name = list(mt.col_key)[0]
+    ck_type = mt.col_key[0].dtype
+    assert ck_type == hl.tstr
+
+    if k and not scores:
+        # column order always preserved?
+        _, scores_table, _ = hwe_normalized_pca(mt, k, compute_loadings=False, as_array=True)
+    elif scores and not k:
+        uid = Env.get_uid()
+        mt = mt.annotate_cols(**{uid: scores[mt.col_key].scores})  # FIXME: check that all cols are annotated
+        scores_table = mt.select_cols(*mt.col_key, scores=mt[uid]).cols()
+    elif k and scores:
+        raise ValueError("pc_relate: exactly one of 'k' and 'scores' must be set, found both")
+    else:
+        raise ValueError("pc_relate: exactly one of 'k' and 'scores' must be set, found neither")
 
-        The fields of the resulting :class:`.Table` entries are of types: `i`:
-        :py:data:`.tstr`, `j`: :py:data:`.tstr`, `kin`: :py:data:`.tfloat64`, `k2`:
-        :py:data:`.tfloat64`, `k1`: :py:data:`.tfloat64`, `k0`:
-        :py:data:`.tfloat64`. The table is keyed by `i` and `j`.
+    scores_local = scores_table.collect()
+    col_key_java = list(map(lambda x: ck_type._convert_to_j(x[ck_name]), scores_local))
```

Comment:
We can't add more convert_to_j calls. We want to delete these methods, not proliferate them.